Broadcast cancellation to only nodes have outstanding child tasks

dnhatn opened the pull request:
Today when canceling a task we broadcast ban/unban requests to all nodes in the cluster. This strategy does not scale well for hierarchical cancellation. With this change, we will track outstanding child requests and broadcast cancellation to only nodes having outstanding child tasks. This change also strengthens the coordination during cancellation. A parent task will no longer be able to send child requests via transport service once it gets cancelled, which streamlines cancellation on the parent task by not requiring to manually check whether the task is cancelled during the execution.

Relates #50990
Supersedes #51157

Co-authored-by: Igor Motov <igor@motovs.org>
Co-authored-by: Yannick Welsch <yannick@welsch.lu>

dnhatn committed:
Propagate cancellation to nodes have outstanding child tasks

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Task Management)

dnhatn commented:
Please hold off the review. I am investigating some test failures that are related to this change.

ywelsch commented:
I wonder if, as part of this iteration, we should already wait with the unbanning until all child tasks have been completed.

dnhatn commented:
> I wonder if, as part of this iteration, we should already wait with the unbanning until all child tasks have been completed (successfully or not).

++. I will add it to this PR.

dnhatn committed:
register child node when execute task locally

dnhatn committed:
Merge branch 'master' into track-child-tasks

dnhatn committed:
only return when childTasks have completed

dnhatn committed:
remove unneed synchronized

dnhatn commented:
This is ready for reviews :).

dnhatn committed:
javadocs

dnhatn committed:
Merge branch 'master' into track-child-tasks

dnhatn committed:
remove assertion

ywelsch commented on server/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -495,6 +552,56 @@ public boolean hasParent(TaskId parentTaskId) {
+                if (childTasksPerNode.isEmpty()) {
+                    runnable = onEmptyChildNodes;
+                } else {
+                    runnable = null;
```
perhaps `runnable = {}`, which avoids the null check later down

dnhatn replied:
`onChildTasksCompleted` is still null if we do not cancel the task.

ywelsch commented on server/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -495,6 +552,56 @@ public boolean hasParent(TaskId parentTaskId) {
+            final Set<DiscoveryNode> pendingChildNodes;
+            synchronized (this) {
+                if (banChildren) {
+                    logger.warn("Trying to start ban twice for task " + task.getId());
```
why does this deserve a warning?

When do we expect this to happen?

ywelsch commented on server/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -495,6 +552,56 @@ public boolean hasParent(TaskId parentTaskId) {
+            synchronized (this) {
+                if (banChildren) {
+                    logger.warn("Trying to start ban twice for task " + task.getId());
+                    pendingChildNodes = Collections.emptySet();
```
Why use empty set here and not the actual set? Will this not lead to incorrectness?

dnhatn replied:
Good catch. Fixed in https://github.com/elastic/elasticsearch/pull/54312/commits/226a5414320e8e4f725a1762e2226af6731a906a

ywelsch commented on server/src/main/java/org/elasticsearch/transport/TransportService.java:
```diff
@@ -560,6 +561,33 @@ public void removeConnectionListener(TransportConnectionListener listener) {
                                                                 TransportResponseHandler<T> handler) {
         try {
+            if (request.getParentTask().isSet()) {
+                final Releasable unregisterChildNode = taskManager.registerChildNode(request.getParentTask().getId(), connection.getNode());
```
I think that we need to actually capture the connection (not the DiscoveryNode object). The reason for this is cross-cluster connections. In the task cancellation action, we assume that the DiscoveryNode is sufficient to resolve the target of the cancellation. This only works for nodes in the same cluster, however.

I don't think that we need to solve this all in this PR, but we need to capture the issue properly, and add an item to the planning.

dnhatn replied:
++ see https://github.com/elastic/elasticsearch/pull/54312/commits/4f337decce08edd355621c9982559df34bc42ce6

ywelsch commented on server/src/main/java/org/elasticsearch/transport/TransportService.java:
```diff
@@ -560,6 +561,33 @@ public void removeConnectionListener(TransportConnectionListener listener) {
+                    @Override
+                    public void handleResponse(T response) {
+                        unregisterChildNode.close();
+                        delegate.handleResponse(response);
```
should we call this in a finally block?

ywelsch commented on server/src/main/java/org/elasticsearch/transport/TransportService.java:
```diff
@@ -560,6 +561,33 @@ public void removeConnectionListener(TransportConnectionListener listener) {
+                    @Override
+                    public void handleException(TransportException exp) {
+                        unregisterChildNode.close();
+                        delegate.handleException(exp);
```
should we call this in a finally block?

ywelsch commented on server/src/test/java/org/elasticsearch/search/SearchCancellationIT.java:
```diff
@@ -151,8 +151,9 @@ public void testCancellationDuringQueryPhase() throws Exception {
         awaitForBlock(plugins);
-        cancelSearch(SearchAction.NAME);
+        ActionFuture<CancelTasksResponse> cancelFuture = cancelSearch(SearchAction.NAME);
         disableBlocks(plugins);
```
assert that `cancelFuture.isDone == false` before disabling blocks?

ywelsch commented on server/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/CancellableTasksIT.java:
```diff
@@ -0,0 +1,381 @@
+        completedLatches.clear();
+    }
+
+    public void testBandOnlyNodesWithOutstandingChildTasks() throws Exception {
```
`testBanOnly...`

ywelsch reviewed (commented):
Thanks Nhat, I've left some questions and comments.

dnhatn committed:
fix tests

dnhatn committed:
fix logging

dnhatn committed:
restore optimization

dnhatn committed:
ban children

dnhatn committed:
add wait_for_completion

dnhatn committed:
Merge branch 'master' into track-child-tasks

dnhatn commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java:
```diff
@@ -104,48 +103,35 @@ protected void processTasks(CancelTasksRequest request, Consumer<CancellableTask
+            final Collection<DiscoveryNode> childNodes =
+                taskManager.startBanOnChildrenNodes(cancellableTask.getId(), () -> groupedListener.onResponse(null));
+            completeListener.whenComplete(r -> {
+                    removeBanOnNodes(cancellableTask, childNodes);
```
Ok, I will try to implement this in a follow-up if it's simple.

dnhatn commented on server/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -163,7 +178,7 @@ private void registerCancellableTask(Task task) {
         CancellableTaskHolder oldHolder = cancellableTasks.put(task.getId(), holder);
         assert oldHolder == null;
         // Check if this task was banned before we start it
-        if (task.getParentTaskId().isSet() && banedParents.isEmpty() == false) {
```
I restored and added a comment https://github.com/elastic/elasticsearch/pull/54312/commits/297b78726aeb4c7cae96db54e98903174745fd0b

dnhatn commented on server/src/main/java/org/elasticsearch/tasks/TaskManager.java:
```diff
@@ -369,6 +397,33 @@ public void removeBan(TaskId parentTaskId) {
+            return holder.startBan(onEmptyChildNodes);
+        } else {
+            logger.warn("Trying to cancel task without registered cancellable task " + taskId);
+            // We still need to set ban on local node for persistent tasks
```
I don't know either. It's from the draft version. Sorry.

dnhatn committed:
Add TODO

dnhatn commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java:
```diff
@@ -104,48 +103,35 @@ protected void processTasks(CancelTasksRequest request, Consumer<CancellableTask
+                    listener.onFailure(e);
                 });
+            canceled = taskManager.cancel(cancellableTask, request.getReason(), () -> groupedListener.onResponse(null));
+            if (canceled) {
```
We throw exception at the end. Anyway, I moved up that statement right after we cancel (see https://github.com/elastic/elasticsearch/pull/54312/commits/a5af596d79d4c663cb711f60f0f1011ad22b32d0).

dnhatn commented:
@ywelsch Thanks for the reviews. I think I've addressed your comments:
- Fix logging issues in https://github.com/elastic/elasticsearch/pull/54312/commits/03ce7a816e1b89712ebd4057d3550646c315af2e
- Add the wait_for_completion parameter and revert test changes in https://github.com/elastic/elasticsearch/pull/54312/commits/a5af596d79d4c663cb711f60f0f1011ad22b32d0

dnhatn committed:
format

dnhatn committed:
wait_for_completion -> wait_for_child_tasks

dnhatn committed:
comment

ywelsch reviewed (changes requested):
Thanks Nhat, I've left some more comments to consider

dnhatn commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java:
```diff
@@ -103,127 +102,75 @@ protected void processTasks(CancelTasksRequest request, Consumer<CancellableTask
+                taskManager.startBanOnChildrenNodes(cancellableTask.getId(), () -> groupedListener.onResponse(null));
+            boolean cancelled = taskManager.cancel(cancellableTask, request.getReason(), () -> groupedListener.onResponse(null));
+            if (cancelled == false) {
+                throw new IllegalStateException("task with id " + cancellableTask.getId() + " is already cancelled");
```
> Another issue I have with this is that this does not play nicely with wait_for_completion. Assume someone else has cancelled the task, and you now want to wait for completion. This gives you a hard exception instead of waiting for all child tasks to complete.

I thought about that scenario but decided not to implement it as I worried about its complexity (we need to keep track of a list of cancellation listeners and childTasksCompleted listeners). Anyway, I will work on it.

ywelsch replied:
Let's add a TODO for that part, and consider it in a follow-up

dnhatn replied:
The `cancel()` method is a bit trappy. I have removed the return value and made it idempotent in https://github.com/elastic/elasticsearch/pull/54312/commits/9a55fe41951f6bf0704ac5fe3690aa1206f832c5.

javanna reviewed (commented):
I left a couple of comments. I am glad to see these improvements made to the tasks framework! 

I think it would be valuable to expand a bit the description of this PR and give it a bit more visibiliry. I find that besides no longer broadcasting bans to all nodes, this PR adds more control to the node where the parent task runs. Today, the parent task has to make sure to stop doing its job once it gets cancelled. With this PR the parent task will no longer be able to send child requests via transport service once it gets cancelled, which streamlines cancellation on the parent task by not requiring to manually check whether the task is cancelled during the execution, ad-hoc This is a great improvement, that search will certainly benefit from.

dnhatn committed:
Allow cancel multiple times

dnhatn committed:
wording

dnhatn committed:
wait_for_completion to false

dnhatn committed:
Merge branch 'master' into track-child-tasks

dnhatn commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java:
```diff
@@ -103,127 +102,75 @@ protected void processTasks(CancelTasksRequest request, Consumer<CancellableTask
+            StepListener<Void> banOnNodesListener = new StepListener<>();
+            setBanOnNodes(request.getReason(), cancellableTask, childrenNodes, banOnNodesListener);
+            banOnNodesListener.whenComplete(groupedListener::onResponse, groupedListener::onFailure);
+            // We remove bans after all child tasks are completed although in theory we can do it on a per-node basic.
```
thanks, fixed in https://github.com/elastic/elasticsearch/pull/54312/commits/a01325e069ff433165a0bdbc0da2854da8db729e

dnhatn commented on rest-api-spec/src/main/resources/rest-api-spec/api/tasks.cancel.json:
```diff
@@ -39,6 +39,10 @@
         "type":"string",
         "description":"Cancel tasks with specified parent task id (node_id:task_number). Set to -1 to cancel all."
+      },
+      "wait_for_child_tasks": {
```
++ I pushed https://github.com/elastic/elasticsearch/pull/54312/commits/8cff3b0a4998ef4f5de015c484c2f4389a807ce3.

dnhatn commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestCancelTasksAction.java:
```diff
@@ -66,6 +66,7 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         cancelTasksRequest.setNodes(nodesIds);
         cancelTasksRequest.setActions(actions);
         cancelTasksRequest.setParentTaskId(parentTaskId);
+        cancelTasksRequest.setWaitForChildTasks(request.paramAsBoolean("wait_for_child_tasks", true));
```
yes, fixed in https://github.com/elastic/elasticsearch/pull/54312/commits/8cff3b0a4998ef4f5de015c484c2f4389a807ce3.

dnhatn commented on x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchActionTests.java:
```diff
@@ -190,8 +193,11 @@ public void testDeleteCancelRunningTask() throws Exception {
-        deleteAsyncSearch(initial.getId());
+        ActionFuture<AcknowledgedResponse> cancelFuture =
+            client().execute(DeleteAsyncSearchAction.INSTANCE, new DeleteAsyncSearchAction.Request(initial.getId()));
+        ensureChildTasksCancelledOrBanned(AsyncSearchId.decode(initial.getId()).getTaskId());
```
Yes, but this is an async call. Anyway, I've reverted this change as wait_for_completion now defaults to false.

dnhatn committed:
allow mock

dnhatn committed:
fix docs

dnhatn commented:
@ywelsch @javanna Thanks for reviewing. This is ready again.

ywelsch commented on client/rest-high-level/src/main/java/org/elasticsearch/client/tasks/CancelTasksRequest.java:
```diff
@@ -138,13 +150,21 @@ public Builder withNodesFiltered(List<String> nodes){
             return this;
         }
 
+        public Builder withWaitForCompletion(boolean waitForCompletion) {
```
Use this in `TasksRequestConvertersTests`?

ywelsch replied:
As well as in `TasksClientDocumentationIT` to document it there as well?

ywelsch commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/CancelTasksRequest.java:
```diff
@@ -68,4 +77,16 @@ public CancelTasksRequest setReason(String reason) {
+        this.waitForCompletion = waitForCompletion;
+    }
+
+    public boolean waitForChildTasks() {
```
call this `getWaitForCompletion`

ywelsch commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestCancelTasksAction.java:
```diff
@@ -66,6 +66,8 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         cancelTasksRequest.setActions(actions);
         cancelTasksRequest.setParentTaskId(parentTaskId);
+        cancelTasksRequest.setWaitForCompletion(
+            request.paramAsBoolean("wait_for_child_tasks", CancelTasksRequest.DEFAULT_WAIT_FOR_COMPLETION));
```
instead of introducing this constant, just use `cancelTasksRequest.getWaitForCompletion()` here

ywelsch reviewed (approved):
I've left a few more nits, no need for another review, however. LGTM

dnhatn committed:
nits

dnhatn commented:
Thanks Yannick.
